crates/primitives/utils/src/service.rs

```diff
@@ -130,142 +528,241 @@ impl Clone for ServiceContext {
+            id: MadaraServiceId::Database,
+            ..Default::default()
         }
     }
 
+    /// Creates a new [Default] [ServiceContext] with the state of its services
+    /// set to the specified value.
+    pub fn new_with_services(services: Arc<MadaraServiceMask>) -> Self {
+        Self { services, ..Default::default() }
+    }
+
     /// Stops all services under the same global context scope.
     pub fn cancel_global(&self) {
+        tracing::info!("🔌 Gracefully shutting down node");
+
         self.token_global.cancel();
     }
 
     /// Stops all services under the same local context scope.
     ///
-    /// A local context is created by calling `branch_local` and allows you to
-    /// reduce the scope of cancellation only to those services which will use
-    /// the new context.
+    /// A local context is created by calling [ServiceContext::child] and allows
+    /// you to reduce the scope of cancellation only to those services which
+    /// will use the new context.
     pub fn cancel_local(&self) {
         self.token_local.as_ref().unwrap_or(&self.token_global).cancel();
     }
 
     /// A future which completes when the service associated to this
     /// [ServiceContext] is canceled.
     ///
-    /// This happens after calling [ServiceContext::cancel_local] or
-    /// [ServiceContext::cancel_global].
+    /// A service is canceled after calling [ServiceContext::cancel_local],
+    /// [ServiceContext::cancel_global] or if it is marked for removal with
+    /// [ServiceContext::service_remove].
     ///
-    /// Use this to race against other futures in a [tokio::select] for example.
+    /// Use this to race against other futures in a [tokio::select] or keep a
+    /// coroutine alive for as long as the service itself.
     #[inline(always)]
-    pub async fn cancelled(&self) {
-        if self.state() != MadaraState::Shutdown {
-            match &self.token_local {
-                Some(token_local) => tokio::select! {
-                    _ = self.token_global.cancelled() => {},
-                    _ = token_local.cancelled() => {}
-                },
-                None => tokio::select! {
-                    _ = self.token_global.cancelled() => {},
-                },
+    pub async fn cancelled(&mut self) {
+        if self.service_update_receiver.is_none() {
+            self.service_update_receiver = Some(self.service_update_sender.subscribe());
+        }
+
+        let mut rx = self.service_update_receiver.take().expect("Receiver was set above");
+        let token_global = &self.token_global;
+        let token_local = self.token_local.as_ref().unwrap_or(&self.token_global);
+
+        loop {
+            // We keep checking for service status updates until a token has
+            // been canceled or this service was deactivated
+            let res = tokio::select! {
+                svc = rx.recv() => svc.ok(),
+                _ = token_global.cancelled() => break,
```

Comment:
why isnt token global a parent of token local? you shouldnt have to await both

Reply:
We have to await both in case the `token_local` is cancelled but the `token_global` is not.